pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/AvoidCallingFinalizeRule.java

```diff
@@ -4,68 +4,75 @@
 
 package net.sourceforge.pmd.lang.java.rule.errorprone;
 
-import java.util.HashSet;
 import java.util.List;
-import java.util.Set;
+import java.util.regex.Pattern;
 
-import net.sourceforge.pmd.lang.java.ast.ASTCompilationUnit;
-import net.sourceforge.pmd.lang.java.ast.ASTName;
+import net.sourceforge.pmd.lang.java.ast.ASTMethodDeclaration;
+import net.sourceforge.pmd.lang.java.ast.ASTPrimaryExpression;
 import net.sourceforge.pmd.lang.java.ast.ASTPrimaryPrefix;
 import net.sourceforge.pmd.lang.java.ast.ASTPrimarySuffix;
+import net.sourceforge.pmd.lang.java.ast.JavaNode;
 import net.sourceforge.pmd.lang.java.rule.AbstractJavaRule;
-import net.sourceforge.pmd.lang.java.symboltable.MethodScope;
-import net.sourceforge.pmd.lang.symboltable.ScopedNode;
 
 public class AvoidCallingFinalizeRule extends AbstractJavaRule {
 
-    private Set<MethodScope> checked = new HashSet<>();
+    private static final Pattern FINALIZE_METHOD_PATTERN = Pattern.compile("^(.+\\.)?finalize$");
 
     @Override
-    public Object visit(ASTCompilationUnit acu, Object ctx) {
-        checked.clear();
-        return super.visit(acu, ctx);
+    public Object visit(ASTPrimaryExpression primaryExpression, Object data) {
+        if (isIncorrectFinalizeMethodCall(primaryExpression)) {
+            addViolation(data, primaryExpression);
+        }
+        return data;
```

Comment:
To be fully correct, we need to call here either `super` or need to use the rule chain. A primary expression could define a anonymous class, which in turn could have primary expressions which call the finalize method....